presto-main/src/main/java/com/facebook/presto/operator/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,188 @@
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        valuesBufferIndex = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        int averageElementSize = Long.BYTES + Byte.BYTES;
+
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += averageElementSize;
+        }
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes)
+    {
+        int averageElementSize = Long.BYTES + Byte.BYTES;
+
+        // it starts from 0 because top level offsets were normalized
+        int lastOffset = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int currentOffset = positionOffsets[i];
+            int entryCount = currentOffset - lastOffset;
+            rowSizes[i] += entryCount * averageElementSize;
+
+            lastOffset = currentOffset;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        verify(decodedBlock != null);
+
+        int[] positions = getPositions();
+        verify(positionsOffset + batchSize - 1 < positions.length && positions[positionsOffset] >= 0 &&
+                positions[positionsOffset + batchSize - 1] < decodedBlock.getPositionCount());
+
+        appendValuesToBuffer();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    @Override
+    public void serializeTo(SliceOutput sliceOutput)
+    {
+        // TODO: getSerializedSizeInBytes() was calculated in flush and doesnt need to recalculated
+        verify(getSerializedSizeInBytes() <= sliceOutput.writableBytes());
+
+        writeLengthPrefixedString(sliceOutput, NAME);
+        sliceOutput.writeInt(bufferedPositionCount);
+        serializeNullsTo(sliceOutput);
+        sliceOutput.appendBytes(valuesBuffer, 0, valuesBufferIndex);
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return getPositionsSizeInBytes() + // positions and mappedPositions
+                valuesBufferIndex +
+                getNullsBufferSizeInBytes();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPostionsRetainedSizeInBytes() +
+                sizeOf(valuesBuffer) +
+                getNullsBufferRetainedSizeInBytes();
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                getNullsBufferSerializedSizeInBytes() +   // nulls buffer
+                valuesBufferIndex;              // values buffer
+    }
+
+    private void appendValuesToBuffer()
+    {
+        verify(decodedBlock != null);
+
+        ensureValueBufferSize();
+
+        if (decodedBlock.mayHaveNull()) {
+            appendLongsWithNullsToBuffer();
```

Comment:
- I'd inline `appendLongsWithNullsToBuffer` and  `appendLongsToBuffer`.
- In `appendLongsWithNullsToBuffer`, it is not necessary to write a (garbage) value if it is null

```
    private void appendValuesToBuffer()
    {
        ensureValuesBufferCapacity();

        int[] positions = getPositions();
        if (decodedBlock.mayHaveNull()) {
            for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
                int position = positions[i];
                if (decodedBlock.isNull(position)) {
                    ByteArrayUtils.writeLong(valuesBuffer, valuesBufferIndex, decodedBlock.getLong(position));
                    valuesBufferIndex += ARRAY_LONG_INDEX_SCALE;
                }
            }
        }
        else {
            for (int i = positionsOffset; i < positionsOffset + batchSize; i++) {
                ByteArrayUtils.writeLong(valuesBuffer, valuesBufferIndex, decodedBlock.getLong(positions[i]));
                valuesBufferIndex += ARRAY_LONG_INDEX_SCALE;
            }
        }
    }
```